### PR TITLE
fix: exit hook

### LINKED ISSUE #2917: Graceful shutdown
## Problem
I'd like to shutdown my application gracefully and do some database work before I exit the process myself.

However, Prisma exits the process on a SIGINT event on its own. Even though I got a work-around by monkey-patching "process.exit", the Prisma query engine still exits with the SIGINT event and I'm unable to execute queries therefore.

`The Node.js process already received a SIGINT signal, therefore the Prisma query engine exited
and your request can't be processed.`

## Suggested solution
1. Make prisma stop listening to process events by setting an option
2. Stop the engines manually by executing a stop method (?) just after the application is ready to exit 

## Alternatives
here's the very hacky work-around that I use (though I'd appreciate something else)

```javascript
const realProcessExit = process.exit;
process.exit = () => {
  // Monkey-patch process.exit because Prisma wants to exit before
  // we can shut down gracefully
};

import { PrismaClient } from "@prisma/client";

["EXIT", "SIGINT", "SIGUSR1", "SIGUSR2", "uncaughtException", "SIGTERM"].forEach((eventType) => {
  const previousListeners = process.listeners(eventType);
  process.removeAllListeners(eventType);

  process.once(eventType, async (param) => {
    await shutdown();

    // Make prisma stop the engines after the application is ready to exit
    previousListeners.forEach((l) => l(param));

    if (eventType !== "EXIT") {
      realProcessExit();
    }
  });
});
```

### LINKED ISSUE #3773: process.on('SIGINT') callback not being called when I import @prisma/client
## The problem
When I import `@prisma/client` my `process.on('SIGINT')` or SIGTERM callbacks are not being called when I use ctrl+c or `kill -15 <pid>`.

## What I think is the root of the problem
I came to the following: when I comment/remove the content of `hookProcess` function here https://github.com/prisma/prisma/blob/ba01d5c2601ad51cdfe7d1602d2f106007e5b749/src/packages/engine-core/src/NodeEngine.ts#L1237 then my process.on('SIGINT') or SIGTERM callbacks do get called. Note: I comment out the body of this function in `node_modules` (/node_modules/@prisma/client/runtime/index.js).

### Bad case (unexpected behavior)
```git clone https://github.com/dmitryrn/patreon-statistics-js patreon-statistics-js-bad-case```
```cd patreon-statistics-js-bad-case```
```git checkout e5e4be7ec2684e250c24511ae94bd746d136fe99```
```npm i```
```npx tsc -p . ```
```DEBUG="*" node cmd/patreon-statistics.js```

Logs:
```log
winston:create-logger: Define prototype method for "error"
winston:create-logger: Define prototype method for "warn"
winston:create-logger: Define prototype method for "info"
winston:create-logger: Define prototype method for "http"
winston:create-logger: Define prototype method for "verbose"
winston:create-logger: Define prototype method for "debug"
winston:create-logger: Define prototype method for "silly"
  prisma-client {
  prisma-client   engineConfig: {
  prisma-client     cwd: '/Users/dmitryrog/projects/patreon-statistics-js/prisma',
  prisma-client     enableDebugLogs: false,
  prisma-client     enableEngineDebugMode: undefined,
  prisma-client     datamodelPath: '/Users/dmitryrog/projects/patreon-statistics-js/node_modules/.prisma/client/schema.prisma',
  prisma-client     prismaPath: undefined,
  prisma-client     engineEndpoint: undefined,
  prisma-client     generator: {
  prisma-client       name: 'client',
  prisma-client       provider: 'prisma-client-js',
  prisma-client       output: '/Users/dmitryrog/projects/patreon-statistics-js/node_modules/@prisma/client',
  prisma-client       binaryTargets: [],
  prisma-client       previewFeatures: [],
  prisma-client       config: {}
  prisma-client     },
  prisma-client     showColors: false,
  prisma-client     logLevel: undefined,
  prisma-client     logQueries: undefined,
  prisma-client     flags: [],
  prisma-client     clientVersion: '2.7.1',
  prisma-client     enableExperimental: [],
  prisma-client     useUds: undefined
  prisma-client   }
  prisma-client } +0ms
winston:create-logger: Define prototype method for "error"
winston:create-logger: Define prototype method for "warn"
winston:create-logger: Define prototype method for "info"
winston:create-logger: Define prototype method for "http"
winston:create-logger: Define prototype method for "verbose"
winston:create-logger: Define prototype method for "debug"
winston:create-logger: Define prototype method for "silly"
{"message":"connected to database","level":"info"}
  express:application set "x-powered-by" to true +0ms
  express:application set "etag" to 'weak' +0ms
  express:application set "etag fn" to [Function: generateETag] +0ms
  express:application set "env" to 'development' +0ms
  express:application set "query parser" to 'extended' +0ms
  express:application set "query parser fn" to [Function: parseExtendedQueryString] +0ms
  express:application set "subdomain offset" to 2 +1ms
  express:application set "trust proxy" to false +0ms
  express:application set "trust proxy fn" to [Function: trustNone] +0ms
  express:application booting in development mode +0ms
  express:application set "view" to [Function: View] +0ms
  express:application set "views" to '/Users/dmitryrog/projects/patreon-statistics-js/views' +0ms
  express:application set "jsonp callback name" to 'callback' +0ms
  express:router use '/' query +1ms
  express:router:layer new '/' +0ms
  express:router use '/' expressInit +0ms
  express:router:layer new '/' +0ms
  express:router:route new '/10-users' +0ms
  express:router:layer new '/10-users' +0ms
  express:router:route get '/10-users' +1ms
  express:router:layer new '/' +0ms
{"port":8080,"level":"info","message":"Server listen on port"}
^C%                 
```

> :warning: **You can see that when I press ctrl+c my process.on callbacks are NOT being called.

### Good case (expected behavior)
```git clone https://github.com/dmitryrn/patreon-statistics-js patreon-statistics-js-good-case```
```cd patreon-statistics-js-good-case```
```git checkout db9a0f596626b195edfd11def5d7e015d34fc16d```
```npm i```
```npx tsc -p .```
```DEBUG="*" node cmd/patreon-statistics.js```

logs:
```log
winston:create-logger: Define prototype method for "error"
winston:create-logger: Define prototype method for "warn"
winston:create-logger: Define prototype method for "info"
winston:create-logger: Define prototype method for "http"
winston:create-logger: Define prototype method for "verbose"
winston:create-logger: Define prototype method for "debug"
winston:create-logger: Define prototype method for "silly"
winston:create-logger: Define prototype method for "error"
winston:create-logger: Define prototype method for "warn"
winston:create-logger: Define prototype method for "info"
winston:create-logger: Define prototype method for "http"
winston:create-logger: Define prototype method for "verbose"
winston:create-logger: Define prototype method for "debug"
winston:create-logger: Define prototype method for "silly"
{"message":"connected to database","level":"info"}
  express:application set "x-powered-by" to true +0ms
  express:application set "etag" to 'weak' +1ms
  express:application set "etag fn" to [Function: generateETag] +0ms
  express:application set "env" to 'development' +0ms
  express:application set "query parser" to 'extended' +1ms
  express:application set "query parser fn" to [Function: parseExtendedQueryString] +0ms
  express:application set "subdomain offset" to 2 +0ms
  express:application set "trust proxy" to false +0ms
  express:application set "trust proxy fn" to [Function: trustNone] +0ms
  express:application booting in development mode +0ms
  express:application set "view" to [Function: View] +0ms
  express:application set "views" to '/Users/dmitryrog/projects/patreon-statistics-js/views' +0ms
  express:application set "jsonp callback name" to 'callback' +0ms
  express:router use '/' query +0ms
  express:router:layer new '/' +1ms
  express:router use '/' expressInit +0ms
  express:router:layer new '/' +0ms
  express:router:route new '/10-users' +0ms
  express:router:layer new '/10-users' +0ms
  express:router:route get '/10-users' +0ms
  express:router:layer new '/' +0ms
{"port":8080,"level":"info","message":"Server listen on port"}
^C{"message":"Received exit signal SIGINT, shutting down gracefully","level":"info"}
{"message":"disconnected from db","level":"info"}
{"message":"Shutting down server","level":"info"}
{"message":"Closed out remaining connections","level":"info"}
{"message":"server stopped","level":"info"}
patreon-statistics exited without error
```

> :warning: **You can see that when I press ctrl+c my callbacks ARE being called.

## My system
MacOS: macOS 11.0 (20A5374i)
Node: v14.9.0
npm/npx: 6.14.8
database used with prisma: postgresql 12
"@prisma/cli": "^2.7.1",
"@prisma/client": "^2.7.1",

### LINKED ISSUE #3917: Stabilize `atomicNumberOperations`
We've fixed remaining the issues and are ready to stabilize this preview feature.

Related: #3491

### PR DESCRIPTION
**Pull request by timsuchanek**
Fix #3773

Fix #2917

### CODE REVIEW

**Review by williamluke4 — APPROVED** [review]

File: src/packages/engine-core/src/NodeEngine.ts [hunk 1]
```diff
@@ -819,12 +832,12 @@ ${this.lastErrorLog.fields.file}:${this.lastErrorLog.fields.line}:${this.lastErr
...
         this.url = `http://localhost:${this.port}`
 
-        // don't wait for this
-        ;(async () => {
-          const engineVersion = await this.version()
-          debug(`Client Version ${this.clientVersion}`)
-          debug(`Engine Version ${engineVersion}`)
-        })()
+          // don't wait for this
+          ; (async () => {
```
Comment (williamluke4) [inline src/packages/engine-core/src/NodeEngine.ts:836]: What is propose of the `;` here?

**Review by Jolg42 — APPROVED** [review]
Looks great!

- I wonder if I should open a candidate issue to test pm2 in e2e tests
- Could you please activate this setting in your VS Code?
Screen Shot 2020-10-14 at 12 46 37

File: src/packages/engine-core/src/NodeEngine.ts (hunk 1 above)
Comment (Jolg42) [reply src/packages/engine-core/src/NodeEngine.ts:836]: It's prettier https://github.com/prettier/prettier/issues/7423

Comment (williamluke4) [reply src/packages/engine-core/src/NodeEngine.ts:836]: Ahh ok, thanks

**Comment by timsuchanek** [comment]
Yep, it's activated 
image

**Comment by timsuchanek** [comment]
Yes, please create one 👍

**Pull request merged**